### ISSUE TITLE
Word wrapping for EditBox

### ISSUE THREAD

**Issue by buggins** [buggins/dlangui#206]

**Comment by joelcnz** [comment]
Yes, I'm needing word wrapping as well!

**Comment by triplejam** [comment]
Hello, I have been implementing this for dlangui. 
Preview in dlangide:
![screenshot_2018-01-07_10-17-45](https://user-images.githubusercontent.com/30878936/34650825-26518b1c-f395-11e7-96fa-278a688f264b.png)

I'm pretty amateur I don't know how much progress has been made on this, but it's kind of a necessity.

Same with selected text: 
![screenshot_2018-01-07_10-16-26](https://user-images.githubusercontent.com/30878936/34650871-e2f12ba6-f395-11e7-938b-774fcc500688.png)

I will make a pull request once I've finished it.

**Comment by and3md** [comment]
@triplejam I tested your algorithm for writing lines and found a few problems. Do you tried to add one very long line? One that will occupy the entire control area, preferably three times or more. Then the vertical scroll bar remains empty and the content can not be scrolled. If you add a pair of lines, you will only be able to jump to the next one. I think you need a cache to compute wrapped content height. 

When I type, the horizontal scroll bar keeps expanding. This happens because its range depends on `_maxLineWidth`. So you probably need to rebuild the function that calculates the maximum length of the line. I try simply set it to `clientRect.width` and `_extendRightScrollBound=false` but this broke selection. 

I did not want to change anything because maybe you're still working on it?

**Comment by triplejam** [comment]
@and3md yes, it's possible there are still certain problems with it. The problems with the horizontal scrollbars may need to manually be set for the program at the moment. As for scrolling down, that's just how dlangui works because of how it counts lines. The wrapped lines are drawn at an offset so they may extend past what's actually counted.  It may help to lower the scrolling speed if it moves too fast. Don't use one long line at the moment, break it up some.

**Comment by triplejam** [comment]
I have some ideas for a "DocumentWidget" which may be better suited for long lines

**Comment by triplejam** [comment]
@and3md Is it working good if you only have a few wrapped lines or if you enable word wrapping for the ircclient, dlangide etc?

**Comment by and3md** [comment]
Yes it works good then. If there is a lot of lines you don't see that scrollbars problem but you never know what user write in editbox :)

**Comment by triplejam** [comment]
@and3md Ok, I've started making some small changes
